BraveShared/Favicons/UIImage+FaviconRenderer.swift

```diff
@@ -0,0 +1,209 @@
+            score -= 1
+          } else {
+            score += 1
+          }
+        }
+        
+        for x in 0..<Int(iconSize.width) {
+          updateScore(x: x, y: 0)
+          updateScore(x: x, y: Int(iconSize.height))
+        }
+
+        // We've already scanned the first and last pixel during
+        // top/bottom pass
+        for y in 1..<Int(iconSize.height) - 1 {
+          updateScore(x: 0, y: y)
+          updateScore(x: Int(iconSize.width), y: y)
+        }
+        return score > 0
+      }
+    }
+    
+    // No alpha channel OR cgImage.dataProvider is nil
+    return false
+  }
+}
+
+// MARK: - Rendering
+extension UIImage {
+  /// Renders an image to a canvas with a background colour and passing
+  static func renderImage(_ image: UIImage, backgroundColor: UIColor?, completion: ((Favicon?) -> Void)?) {
+    if let cgImage = image.cgImage, image.hasTransparentEdges {
+      let padding = 4.0
+      let size = CGSize(
+        width: image.size.width + padding,
+        height: image.size.height + padding)
+
+      let finalImage = drawOnImageContext(size: size) { context, rect in
+        context.saveGState()
+        context.setFillColor(backgroundColor?.cgColor ?? UIColor.clear.cgColor)
+        context.fill(rect)
+
+        context.translateBy(x: 0.0, y: rect.size.height)
+        context.scaleBy(x: 1.0, y: -1.0)
+
+        context.draw(cgImage, in: rect.insetBy(dx: padding, dy: padding))
+        context.restoreGState()
+      }
+      completion?(Favicon(image: finalImage ?? image, isMonogramImage: false, backgroundColor: backgroundColor ?? .clear))
+    } else {
+      completion?(Favicon(image: image, isMonogramImage: false, backgroundColor: backgroundColor ?? .clear))
+    }
+  }
+  
+  /// Renders a monogram letter of the specified URL onto a canvas with textColor and backgroundColor
+  static func renderMonogram(_ url: URL, textColor: UIColor?, backgroundColor: UIColor?, monogramString: String?, completion: ((Favicon?) -> Void)?) {
+    // Monogram favicon attributes
+    let createBackgroundColor = { (url: URL) -> UIColor in
+      guard let hash = url.baseDomain?.hashValue else {
+        return .gray
+      }
+      let index = abs(hash) % (UIConstants.defaultColorStrings.count - 1)
+      let colorHex = UIConstants.defaultColorStrings[index]
+      return UIColor(colorString: colorHex)
+    }
+    
+    let solidTextColor = textColor ?? .white
+    let solidBackgroundColor = backgroundColor ?? createBackgroundColor(url)
+    
+    let label = UILabel()
+    label.textColor = solidTextColor
+    label.backgroundColor = solidBackgroundColor
+    label.minimumScaleFactor = 0.5
+
+    let text = (monogramString ?? FaviconUtils.monogramLetter(
+        for: url,
+        fallbackCharacter: nil
+      )) as NSString
+
+    let padding = 4.0
+    let finalImage = drawOnImageContext { context, rect in
+      guard let font = label.font else { return }
+      var fontSize = font.pointSize
```

Comment:
How will this work with dynamic font? `UILabel`'s font by default is going to be the preferred body font which means it will grow as the users font size grows. Have you tested the monogram code with multiple font sizes?